README.md
<img src="https://rawgit.com/Spirals-Team/powerapi/master/resources/logo/PowerAPI-logo.png" alt="Powerapi" width="300px">

[![Join the chat at https://gitter.im/Spirals-Team/powerapi](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/Spirals-Team/powerapi?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

[![License: AGPL v3](https://img.shields.io/badge/License-AGPL%20v3-blue.svg)](http://www.gnu.org/licenses/agpl-3.0)
[![Build Status](https://travis-ci.org/Spirals-Team/powerapi.svg?branch=master)](https://travis-ci.org/Spirals-Team/powerapi)
[![Coverage Status](https://coveralls.io/repos/Spirals-Team/powerapi/badge.svg)](https://coveralls.io/r/Spirals-Team/powerapi)

PowerAPI is a middleware toolkit for building software-defined power meters.
Software-defined power meters are configurable software libraries that can estimate the power consumption of software in real-time.
PowerAPI supports the acquisition of raw metrics from a wide diversity of sensors (*eg.*, physical meters, processor interfaces, hardware counters, OS counters) and the delivery of power consumptions via different channels (including file system, network, web, graphical).
As a middleware toolkit, PowerAPI offers the capability of assembling power meters *«à la carte»* to accommodate user requirements.

# About

PowerAPI is an open-source project developed by the [Spirals research group](https://team.inria.fr/spirals) (University of Lille 1 and Inria) and fully managed with [sbt](http://www.scala-sbt.org).

The documentation is available [here](https://github.com/Spirals-Team/powerapi/wiki/Getting-started).

## Mailing list
You can follow the latest news and asks questions by subscribing to our <a href="mailto:sympa@inria.fr?subject=subscribe powerapi">mailing list</a>.

## Contributing
If you would like to contribute code you can do so through GitHub by forking the repository and sending a pull request.

When submitting code, please make every effort to follow existing conventions and style in order to keep the code as readable as possible.

## Publications
* **[WattsKit: Software-Defined Power Monitoring of Distributed Systems](https://hal.inria.fr/hal-01439889)**: M. Colmant, P. Felber, R. Rouvoy, L. Seinturier. *IEEE/ACM International Symposium on Cluster, Cloud and Grid Computing* (CCGrid). April 2017, Spain, France. pp.1-14.
* **[Process-level Power Estimation in VM-based Systems](https://hal.inria.fr/hal-01130030)**: M. Colmant, M. Kurpicz, L. Huertas, R. Rouvoy, P. Felber, A. Sobe. *European Conference on Computer Systems* (EuroSys). April 2015, Bordeaux, France. pp.1-14.
* **[Monitoring Energy Hotspots in Software](https://hal.inria.fr/hal-01069142)**: A. Noureddine, R. Rouvoy, L. Seinturier. *Journal of Automated Software Engineering*, Springer, 2015, pp.1-42.
* **[Unit Testing of Energy Consumption of Software Libraries](https://hal.inria.fr/hal-00912613)**: A. Noureddine, R. Rouvoy, L. Seinturier. *International Symposium On Applied Computing* (SAC), March 2014, Gyeongju, South Korea. pp.1200-1205.
* **[Informatique : Des logiciels mis au vert](http://www.jinnove.com/Actualites/Informatique-des-logiciels-mis-au-vert)**: L. Seinturier, R. Rouvoy. *J'innove en Nord Pas de Calais*, [NFID](http://www.jinnove.com), 2013.
* **[PowerAPI: A Software Library to Monitor the Energy Consumed at the Process-Level](http://ercim-news.ercim.eu/en92/special/powerapi-a-software-library-to-monitor-the-energy-consumed-at-the-process-level)**: A. Bourdon, A. Noureddine, R. Rouvoy, L. Seinturier. *ERCIM News, Special Theme: Smart Energy Systems*, 92,  pp.43-44. [ERCIM](http://www.ercim.eu), 2013.
* **[Mesurer la consommation en énergie des logiciels avec précision](http://www.lifl.fr/digitalAssets/0/807_01info_130110_16_39.pdf)**: A. Bourdon, R. Rouvoy, L. Seinturier. *01 Business & Technologies*, 2013.
* **[A review of energy measurement approaches](https://hal.inria.fr/hal-00912996v2)**: A. Noureddine, R. Rouvoy, L. Seinturier. *ACM SIGOPS Operating Systems Review*, ACM, 2013, 47 (3), pp.42-49.
* **[Runtime Monitoring of Software Energy Hotspots](https://hal.inria.fr/hal-00715331)**: A. Noureddine, A. Bourdon, R. Rouvoy, L. Seinturier. *International Conference on Automated Software Engineering* (ASE), September 2012, Essen, Germany. pp.160-169.
* **[A Preliminary Study of the Impact of Software Engineering on GreenIT](https://hal.inria.fr/hal-00681560)**: A. Noureddine, A. Bourdon, R. Rouvoy, L. Seinturier. *International Workshop on Green and Sustainable Software* (GREENS), June 2012, Zurich, Switzerland. pp.21-27.

## Use Cases
PowerAPI is used in a variety of projects to address key challenges of GreenIT:
* [GenPack](https://hal.inria.fr/hal-01403486) provides a Docker Swarm strategy to minimize the energy footprint of  Docker containers deployed in a cluster
* [BitWatts](http://bitwatts.powerapi.org) provides process-level power estimation of applications running in virtual machines
* [Web Energy Archive](http://webenergyarchive.com) ranks popular websites based on the energy footpring they imposes to browsers
* [Greenspector](http://greenspector.com) optimises the power consumption of software by identifying potential energy leaks in the source code.

## Acknowledgments
We all stand on the shoulders of giants and get by with a little help from our friends. PowerAPI is written in [Scala](http://www.scala-lang.org) (version 2.12.1 under [3-clause BSD license](http://www.scala-lang.org/license.html)) and built on top of:
* [Akka](http://akka.io) (version 2.4.14 under [Apache 2 license](http://www.apache.org/licenses/LICENSE-2.0)), for asynchronous processing.
* [Typesafe Config](https://github.com/typesafehub/config) (version 1.3.1 under [Apache 2 license](http://www.apache.org/licenses/LICENSE-2.0)), for reading configuration files.
* [scala-logging](https://github.com/typesafehub/scala-logging) (version 3.5.0 under [Apache 2 license](http://www.apache.org/licenses/LICENSE-2.0)), for Scala wrapping SL4J.
* [logback](https://github.com/qos-ch/logback) (version 1.1.7 under [LGPL 2.1 license](https://github.com/qos-ch/logback/blob/master/LICENSE.txt)), for logging purpose.
* [powerspy.scala](https://github.com/Spirals-Team/powerspy.scala) (version 1.2 under [AGPL license](http://www.gnu.org/licenses/agpl-3.0.html)), for using the [PowerSpy](http://www.alciom.com/en/products/powerspy2-en-gb-2.html) power meter.
* [BridJ](https://code.google.com/p/bridj/) (version 0.7.0 under [3-clause BSD license](https://github.com/ochafik/nativelibs4java/blob/master/libraries/BridJ/LICENSE)), for system or C calls.
* [JNA](https://github.com/twall/jna) (version 4.4.0 under [LGPL 2.1 license](https://github.com/twall/jna/blob/master/LGPL2.1)), for system or C calls.
* [perfmon2](http://sourceforge.net/p/perfmon2/libpfm4/ci/master/tree) (version 4.7.0 under [MIT license](http://sourceforge.net/p/perfmon2/libpfm4/ci/master/tree/COPYING)), for accessing hardware performance counters.
* [JFreeChart](http://www.jfree.org/jfreechart/) (version 1.0.19 under [LGPL license](https://www.gnu.org/licenses/lgpl.html)), for creation of interactive and animated charts.
* [grizzled-scala](http://software.clapper.org/grizzled-scala/) (version 4.0.0 under [3-clause BSD license](https://github.com/bmc/grizzled-scala/blob/master/LICENSE.md)), for new utility classes and objects.
* [Sigar](https://support.hyperic.com/display/SIGAR/Home) (version 1.6.5 under [Apache 2 license](http://www.apache.org/licenses/LICENSE-2.0)), for providing a portable interface for gathering system information.
* [spray-json](http://spray.io/) (version 1.3.2 under [Apache 2 license](http://www.apache.org/licenses/LICENSE-2.0)), for (de)serializing JSON.
* [scala-influxdb-client](https://github.com/paulgoldbaum/scala-influxdb-client) (version 0.5.2 under [MIT license](https://github.com/paulgoldbaum/scala-influxdb-client/blob/master/LICENSE)), for using an asynchronous scala API for InfluxDB.

# License
This software is licensed under the *GNU Affero General Public License*, quoted below.

Copyright (C) 2011-2017 Inria, University of Lille 1.

PowerAPI is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

PowerAPI is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License along with PowerAPI. If not, please consult http://www.gnu.org/licenses/agpl-3.0.html.


CHANGELOG.md
# Change Log

## [4.2.2](https://github.com/Spirals-Team/powerapi/tree/4.2.2) (2018-03-02)
[Full Changelog](https://github.com/Spirals-Team/powerapi/compare/4.2.1...4.2.2)

**Implemented enhancements:**

- The containers can now be monitored using theirs Name, short and full ID.
- Use of the container's name as target name instead of the full ID for the power measurements reporting.
- Trying to monitor an unknown container is considered as fatal error and PowerAPI will shutdown.
- Verify Docker socket availability when containers are monitored.
- Use the same working directory (/powerapi) for cli and cpu sampling Docker images.

## [4.2.1](https://github.com/Spirals-Team/powerapi/tree/4.2.1) (2018-02-19)
[Full Changelog](https://github.com/Spirals-Team/powerapi/compare/4.2...4.2.1)

**Implemented enhancements:**

- Rework of PowerAPI Docker images : based on Ubuntu Xenial (LTS), use of distribution packages instead of building the dependencies from sources

**Fixed bugs:**

- CPU sampling: Add filter to CPU device(s) selection
- RAPL: Remove MSR kernel module load (problems in Docker images)
- Travis CI build: Docker images are now properly pushed to Docker hub

## [4.2](https://github.com/Spirals-Team/powerapi/tree/4.2) (2017-11-02)
[Full Changelog](https://github.com/Spirals-Team/powerapi/compare/4.0...4.2)

**Implemented enhancements:**

- Update projects and libraries to Scala version 2.12
- Implement a disk module (only for Linux)
- Use an asynchronous API for InfluxDB
- Use the cgroup statistics to retrieve the cpu usage of a docker container
- Improving the resilience of power conversions

**Fixed bugs:**

- Bluecove libraries download URL for sbt
- Travis CI build: JDK version 7 to 8, Ruby installation, build failing on pull-request

**Closed issues:**

- Version mismatch [\#87](https://github.com/Spirals-Team/powerapi/issues/87)

## [4.0](https://github.com/Spirals-Team/powerapi/tree/4.0) (2016-04-14)
[Full Changelog](https://github.com/Spirals-Team/powerapi/compare/3.4...4.0)

**Implemented enhancements:**

- Use a settings file to declare the architectures compatible with RAPL [\#78](https://github.com/Spirals-Team/powerapi/issues/78)
- Packaging PowerAPI as docker images [\#77](https://github.com/Spirals-Team/powerapi/issues/77)
- feature\(v4.0\): refactoring [\#81](https://github.com/Spirals-Team/powerapi/pull/81) ([mcolmant](https://github.com/mcolmant))
- feature\(v4.0\): refactoring [\#80](https://github.com/Spirals-Team/powerapi/pull/80) ([mcolmant](https://github.com/mcolmant))

**Fixed bugs:**

- Cannot compile CLI: “error: not found: value Universal” [\#79](https://github.com/Spirals-Team/powerapi/issues/79)

## [3.4](https://github.com/Spirals-Team/powerapi/tree/3.4) (2016-02-02)
[Full Changelog](https://github.com/Spirals-Team/powerapi/compare/3.3...3.4)

**Closed issues:**

- Link error while using Sigar module [\#75](https://github.com/Spirals-Team/powerapi/issues/75)

## [3.3](https://github.com/Spirals-Team/powerapi/tree/3.3) (2015-11-03)
[Full Changelog](https://github.com/Spirals-Team/powerapi/compare/3.2...3.3)

**Implemented enhancements:**

- Refactors the PowerSpy module [\#53](https://github.com/Spirals-Team/powerapi/issues/53)
- Create the wiki with the documentation [\#46](https://github.com/Spirals-Team/powerapi/issues/46)
- Implement the filesystem interface [\#16](https://github.com/Spirals-Team/powerapi/issues/16)
- Implement the daemon mode [\#15](https://github.com/Spirals-Team/powerapi/issues/15)
- Implement the Web interface [\#14](https://github.com/Spirals-Team/powerapi/issues/14)

**Merged pull requests:**

- feature\(docker\): implement energy monitoring of a container [\#74](https://github.com/Spirals-Team/powerapi/pull/74) ([huertas](https://github.com/huertas))
- feature\(g5k-sensor\): implement the external powermeter OmegaWatt \(fro… [\#73](https://github.com/Spirals-Team/powerapi/pull/73) ([huertas](https://github.com/huertas))
- feature\(rest\): implement the REST reporter [\#72](https://github.com/Spirals-Team/powerapi/pull/72) ([huertas](https://github.com/huertas))
- feature\(fuse\): implement the FUSE reporter [\#71](https://github.com/Spirals-Team/powerapi/pull/71) ([huertas](https://github.com/huertas))
- feature\(daemon\): implement the PowerAPI daemon [\#70](https://github.com/Spirals-Team/powerapi/pull/70) ([huertas](https://github.com/huertas))

## [3.2](https://github.com/Spirals-Team/powerapi/tree/3.2) (2015-06-02)
[Full Changelog](https://github.com/Spirals-Team/powerapi/compare/3.1...3.2)

**Implemented enhancements:**

- Give the possibility to give directly the parameters to the PowerModule [\#60](https://github.com/Spirals-Team/powerapi/issues/60)
- Increase the actors timeout [\#59](https://github.com/Spirals-Team/powerapi/issues/59)
- Add prefix for each parameter written in configuration files [\#58](https://github.com/Spirals-Team/powerapi/issues/58)
- Adds the MUID as an argument in the PowerDisplay trait [\#56](https://github.com/Spirals-Team/powerapi/issues/56)
- Stops actors created for each PowerDisplay when a Monitor is canceled [\#55](https://github.com/Spirals-Team/powerapi/issues/55)
- Refactors the LibpfmChildSensor [\#54](https://github.com/Spirals-Team/powerapi/issues/54)
- Fix estimations returned by Sigar module under Windows [\#61](https://github.com/Spirals-Team/powerapi/issues/61)
- Implement the SIGAR module [\#17](https://github.com/Spirals-Team/powerapi/issues/17)
- Merge branch "develop" into "master" [\#69](https://github.com/Spirals-Team/powerapi/pull/69) ([mcolmant](https://github.com/mcolmant))
- fix\(\#58\): Refactors the Configuration trait [\#67](https://github.com/Spirals-Team/powerapi/pull/67) ([mcolmant](https://github.com/mcolmant))
- fix\(\#54\): Refactors the LibpfmChildSensor [\#66](https://github.com/Spirals-Team/powerapi/pull/66) ([mcolmant](https://github.com/mcolmant))

docker/agent/payload.pb.c
/*
 * Generated by the protobluff compiler - do not edit.
 * ----------------------------------------------------------------------------
 * Date:     03/29/16 14:06:32
 * Filename: payload.proto
 * Package:  -
 * ----------------------------------------------------------------------------
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

#include <protobluff/descriptor.h>

#include "payload.pb.h"

/* ----------------------------------------------------------------------------
 * Descriptors
 * ------------------------------------------------------------------------- */

/* MapEntry : descriptor */
pb_descriptor_t
mapentry_descriptor = { {
  (const pb_field_descriptor_t []){

    /* required string key = 1 */
    { .tag   = 1,
      .name  = "key",
      .type  = STRING,
      .label = REQUIRED },

    /* required uint64 value = 2 */
    { .tag   = 2,
      .name  = "value",
      .type  = UINT64,
      .label = REQUIRED }

  }, 2 } };

/* Payload : descriptor */
pb_descriptor_t
payload_descriptor = { {
  (const pb_field_descriptor_t []){

    /* required uint32 core = 1 */
    { .tag   = 1,
      .name  = "core",
      .type  = UINT32,
      .label = REQUIRED },

    /* required uint32 pid = 2 */
    { .tag   = 2,
      .name  = "pid",
      .type  = UINT32,
      .label = REQUIRED },

    /* required uint32 tid = 3 */
    { .tag   = 3,
      .name  = "tid",
      .type  = UINT32,
      .label = REQUIRED },

    /* required uint64 timestamp = 4 */
    { .tag   = 4,
      .name  = "timestamp",
      .type  = UINT64,
      .label = REQUIRED },

    /* repeated MapEntry counters = 5 */
    { .tag   = 5,
      .name  = "counters",
      .type  = MESSAGE,
      .label = REPEATED,
      .refer = &mapentry_descriptor },

    /* repeated string traces = 6 */
    { .tag   = 6,
      .name  = "traces",
      .type  = STRING,
      .label = REPEATED }

  }, 6 } };

